src/tracebox/tracebox_ctl.cc

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/tracebox/tracebox_ctl.h"
+
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <cerrno>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <string>
+#include <vector>
+
+#include "perfetto/base/build_config.h"
+#include "perfetto/base/logging.h"
+#include "perfetto/base/time.h"
+#include "perfetto/ext/base/file_utils.h"
+#include "perfetto/ext/base/string_utils.h"
+#include "perfetto/ext/base/subprocess.h"
+#include "perfetto/ext/base/temp_file.h"
+#include "perfetto/ext/base/unix_socket.h"
+#include "perfetto/ext/base/utils.h"
+#include "perfetto/tracing/default_socket.h"
+
+#if !PERFETTO_BUILDFLAG(PERFETTO_OS_WIN)
+#include <sys/wait.h>
+#include <unistd.h>
+#endif
+
+namespace perfetto {
+namespace {
+
+constexpr const char* kDaemons[] = {
+    "traced",
+    "traced_probes",
+#if PERFETTO_BUILDFLAG(PERFETTO_TRACED_PERF)
+    "traced_perf",
+#endif
+};
+
+bool CanConnectToSocket(const std::string& path) {
+  base::SockFamily family = base::GetSockFamily(path.c_str());
+  auto sock =
+      base::UnixSocketRaw::CreateMayFail(family, base::SockType::kStream);
+  return sock && sock.Connect(path);
+}
+
+#if !PERFETTO_BUILDFLAG(PERFETTO_OS_WIN)
+
+std::string GetPidFilePath(const char* daemon_name) {
+  return base::GetSysTempDir() + "/" + daemon_name + ".pid";
+}
+
+bool WritePidToFile(const std::string& path, pid_t pid) {
+  base::ScopedFile fd =
+      base::OpenFile(path, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+  if (!fd)
+    return false;
+  std::string content = std::to_string(pid);
+  auto written = base::WriteAll(fd.get(), content.c_str(), content.size());
+  return written >= 0 && static_cast<size_t>(written) == content.size();
+}
+
+pid_t ReadPidFromFile(const std::string& path) {
+  std::string pid_str;
+  if (!base::ReadFile(path, &pid_str))
+    return 0;
+  pid_str = base::TrimWhitespace(pid_str);
+  if (pid_str.empty())
+    return 0;
```

Comment:
I think this case (empty) is already covered below by StringTOInt64 (check, but I think it fails if it's empty)